# **Asignación: Modelo de Regresión con Keras**

## **1. Tema de la asignación**

En este proyecto, construiremos un modelo de **regresión** utilizando la biblioteca **Keras** para modelar los datos sobre la **resistencia a la compresión del hormigón**. Este ejercicio es una continuación del análisis realizado en el **Laboratorio 3**, aplicando redes neuronales artificiales para predecir la resistencia del material en función de sus componentes.

---

## **2. Datos del hormigón**

Para su comodidad, los datos se pueden encontrar en el siguiente enlace:

🔗 [Descargar datos](https://cocl.us/concrete_data)

Los datos contienen información sobre la resistencia del hormigón en función de los siguientes predictores:

- **Cemento**
- **Escoria de alto horno**
- **Cenizas volantes**
- **Agua**
- **Superplastificante**
- **Árido grueso**
- **Árido fino**

El objetivo es construir un modelo que prediga la **resistencia a la compresión** del hormigón basándose en estos factores.

---

## **3. Instrucciones para la tarea**

1. **Carga de datos**: Importar y visualizar el conjunto de datos para entender su estructura.
2. **Preprocesamiento**: Normalización y división en conjuntos de entrenamiento y prueba.
3. **Construcción del modelo**: Implementación de una red neuronal con Keras para regresión.
4. **Entrenamiento y evaluación**: Comparar el rendimiento del modelo a través de métricas adecuadas, como el error medio cuadrático (*MSE*).
5. **Discusión de resultados**: Comparar la diferencia en la media de los errores obtenidos entre distintas configuraciones del modelo.

---


## **4. Código de implementación**

In [1]:
# Importar bibliotecas necesarias

# pandas: Librería para manipulación y análisis de datos estructurados (DataFrames y Series)
import pandas as pd

# numpy: Librería para operaciones numéricas y manejo de matrices/arreglos
import numpy as np

# tensorflow: Marco de trabajo para el desarrollo de modelos de aprendizaje profundo
import tensorflow as tf

# keras: API de alto nivel dentro de TensorFlow para construir y entrenar redes neuronales
from tensorflow import keras

# Sequential: Clase de Keras que permite construir modelos de redes neuronales capa por capa
from tensorflow.keras.models import Sequential

# Dense: Capa de neuronas totalmente conectada utilizada en redes neuronales artificiales
from tensorflow.keras.layers import Dense

# train_test_split: Función de Scikit-learn para dividir los datos en conjuntos de entrenamiento y prueba
from sklearn.model_selection import train_test_split

# StandardScaler: Método para normalizar los datos asegurando que tengan media 0 y desviación estándar 1
from sklearn.preprocessing import StandardScaler

# mean_squared_error: Métrica para evaluar la precisión del modelo comparando valores predichos con los reales
from sklearn.metrics import mean_squared_error

In [7]:
# Importar la función Input de Keras para definir explícitamente la capa de entrada
from tensorflow.keras import Input

In [2]:
# Cargar los datos
data_url = "https://cocl.us/concrete_data"
df = pd.read_csv(data_url)

In [3]:
# Separar variables predictoras y variable objetivo
X = df.drop(columns=['Strength'])  # Variables de entrada
y = df['Strength']  # Variable objetivo

In [12]:
# Inicializar lista para almacenar los errores
mse_list = []

In [14]:
# Repetir el proceso 50 veces
for _ in range(50):
    # Dividir los datos en conjunto de entrenamiento (70%) y prueba (30%)
    X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=None)

    # Normalizar los datos
    scaler = StandardScaler()
    X_train = scaler.fit_transform(X_train)
    X_test = scaler.transform(X_test)

    # Construcción del modelo con la recomendación de Keras
    model = Sequential([
        Input(shape=(X_train.shape[1],)),  # Definir la capa de entrada explícitamente
        Dense(64, activation='relu'),  # Primera capa oculta con 64 neuronas y ReLU
        Dense(64, activation='relu'),  # Segunda capa oculta con 64 neuronas y ReLU
        Dense(1)  # Capa de salida con 1 neurona para regresión
    ])

    # Compilar el modelo
    model.compile(optimizer='adam', loss='mse')

    # Entrenar el modelo con 50 épocas
    model.fit(X_train, y_train, epochs=50, verbose=0, batch_size=10)

    # Evaluación del modelo
    y_pred = model.predict(X_test)
    mse = mean_squared_error(y_test, y_pred)
    mse_list.append(mse)

10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 7ms/step 
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 18ms/step
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 7ms/step 
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 11ms/step
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 7ms/step 
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 7ms/step 
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 7ms/step 
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 7ms/step 
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 10ms/step
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 10ms/step
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 7ms/step 
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 12ms/step
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 7ms/step 
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 7ms/step 
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 8ms/step 
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 7ms/step 
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 12ms/step
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 12ms/step
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 7ms/step 
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 7ms/step 
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 7ms/step 
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 9ms/step
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 7ms/step 
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 10ms/step
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 9ms/step
10

In [15]:
# Calcular la media y desviación estándar de los errores
mse_mean = np.mean(mse_list)
mse_std = np.std(mse_list)

In [16]:
print(f'Error Medio Cuadrático promedio: {mse_mean:.4f}')
print(f'Desviación estándar del MSE: {mse_std:.4f}')

Error Medio Cuadrático promedio: 49.6710
Desviación estándar del MSE: 25.4673


## Conclusión

En este experimento, se construyó un modelo de red neuronal para predecir la resistencia a la compresión del hormigón a partir de sus componentes. Tras realizar 50 repeticiones del proceso de entrenamiento y evaluación, se obtuvo un error medio cuadrático (MSE) promedio de **49.6710** con una desviación estándar de **25.4673**.

Estos resultados indican que el modelo tiene un desempeño estable; sin embargo, la alta desviación estándar sugiere que la precisión de las predicciones puede verse afectada por la variabilidad en la selección del conjunto de entrenamiento. Para mejorar su rendimiento, podrían implementarse estrategias como el ajuste de hiperparámetros, el uso de arquitecturas más complejas o el aumento del tamaño del conjunto de datos.

La red neuronal construida logra modelar la relación entre los componentes del hormigón y su resistencia, pero aún presenta margen de mejora para optimizar su precisión y reducir la variabilidad en sus predicciones.
